File: algorithm/leetcodeObjectiveC/combination-sum-III.m
// 216. Combination Sum III
// Difficulty: Medium

// Find all possible combinations of k numbers that add up to a number n,
// given that only numbers from 1 to 9 can be used and each combination should be a unique set of numbers.


// Example 1:
// Input: k = 3, n = 7

// Output:
// [[1,2,4]]

// Example 2:
// Input: k = 3, n = 9

// Output:
// [[1,2,6], [1,3,5], [2,3,4]]

// Time:  O(k * C(n, k))
// Space: O(k)

#import <Foundation/Foundation.h>

void combinationSum3DFS(int k, int n, int level, NSMutableArray** path, NSMutableArray** res) {
  if (n < 0) {
    return;
  }
  if (n == 0 && [*path count] == k) {
    [*res addObject:[*path copy]];
  }
  for (int i = level; i <= 9; ++i) {
    [*path addObject: @(i)];
    combinationSum3DFS(k, n - i, i + 1, path, res);
    [*path removeLastObject];
  }
}

NSArray* combinationSum3(int k, int n) {
  NSMutableArray* res = @[].mutableCopy;
  NSMutableArray* path = @[].mutableCopy;
  combinationSum3DFS(k, n, 1, &path, &res);
  return res;
}

File: data-structure-and-algorithm/javascript/leetcode/combination-sum-iii.js
/*
Find all valid combinations of k numbers that sum up to n such that the following conditions are true:

Only numbers 1 through 9 are used.
Each number is used at most once.
Return a list of all possible valid combinations. The list must not contain the same combination twice, and the combinations may be returned in any order.

 

Example 1:

Input: k = 3, n = 7
Output: [[1,2,4]]
Explanation:
1 + 2 + 4 = 7
There are no other valid combinations.
Example 2:

Input: k = 3, n = 9
Output: [[1,2,6],[1,3,5],[2,3,4]]
Explanation:
1 + 2 + 6 = 9
1 + 3 + 5 = 9
2 + 3 + 4 = 9
There are no other valid combinations.
Example 3:

Input: k = 4, n = 1
Output: []
Explanation: There are no valid combinations.
Using 4 different numbers in the range [1,9], the smallest sum we can get is 1+2+3+4 = 10 and since 10 > 1, there are no valid combination.
Example 4:

Input: k = 3, n = 2
Output: []
Explanation: There are no valid combinations.
Example 5:

Input: k = 9, n = 45
Output: [[1,2,3,4,5,6,7,8,9]]
Explanation:
1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9 = 45
There are no other valid combinations.
 

Constraints:

2 <= k <= 9
1 <= n <= 60
*/

function combinationSum3(k, n) {
    const result = [];
    dfs(n, k, 1, [], result);
    return result;
}

function dfs( target, count, left, curr, result) {
    if (target === 0 && curr.length === count) {
        result.push([...curr]);
        return;
    }
    for (let i = left; i <= 9; ++i) {
        const diff = target - i;
        if (diff >= 0 && curr.length < count) {
            curr.push(i);
            dfs(diff, count, i + 1, curr, result);
            curr.pop();
        }
    }
}